File: web/src/climbPerformance.ts
import type { ActivityClimb, ActivityLap, ActivitySample } from "./types";

export type ClimbPerformance = {
  paceSPKM?: number;
  gapSPKM?: number;
};

export function samplesForClimbPerformance(fullSamples?: ActivitySample[], seriesSamples?: ActivitySample[]): ActivitySample[] {
  if (fullSamples && fullSamples.length > 0) {
    return fullSamples;
  }
  return seriesSamples ?? [];
}

export function climbPerformanceFor(samples: ActivitySample[], laps: ActivityLap[], climb: ActivityClimb): ClimbPerformance {
  return {
    paceSPKM: paceForClimbSamples(samples, climb),
    gapSPKM: gapForClimbLaps(laps, climb)
  };
}

export function paceForClimbSamples(samples: ActivitySample[], climb: ActivityClimb): number | undefined {
  const climbSamples = samplesForClimb(samples, climb);
  if (climbSamples.length < 2) {
    return undefined;
  }

  let distanceM = 0;
  let movingTimeS = 0;
  for (let index = 1; index < climbSamples.length; index += 1) {
    const previous = climbSamples[index - 1];
    const current = climbSamples[index];
    const previousDistance = finiteDistance(previous.distanceM);
    const currentDistance = finiteDistance(current.distanceM);
    if (previousDistance === undefined || currentDistance === undefined) {
      continue;
    }
    const distanceDeltaM = currentDistance - previousDistance;
    if (distanceDeltaM <= 0) {
      continue;
    }

    const durationS = sampleDurationSeconds(previous, current, distanceDeltaM);
    if (durationS === undefined || durationS <= 0) {
      continue;
    }
    distanceM += distanceDeltaM;
    movingTimeS += durationS;
  }

  if (distanceM <= 0 || movingTimeS <= 0) {
    return undefined;
  }
  return movingTimeS / distanceM * 1000;
}

export function gapForClimbLaps(laps: ActivityLap[], climb: ActivityClimb): number | undefined {
  const startDistanceM = Math.min(climb.startDistanceM, climb.endDistanceM);
  const endDistanceM = Math.max(climb.startDistanceM, climb.endDistanceM);
  let lapStartDistanceM = 0;
  let weightedGap = 0;
  let weightedDistance = 0;

  for (const lap of laps.slice().sort((left, right) => left.index - right.index)) {
    const lapDistanceM = finitePositiveOrZero(lap.distanceM);
    const lapEndDistanceM = lapStartDistanceM + lapDistanceM;
    const overlapDistanceM = Math.max(0, Math.min(endDistanceM, lapEndDistanceM) - Math.max(startDistanceM, lapStartDistanceM));
    if (overlapDistanceM > 0 && validPace(lap.avgGradeAdjustedPaceSPKM)) {
      weightedGap += lap.avgGradeAdjustedPaceSPKM! * overlapDistanceM;
      weightedDistance += overlapDistanceM;
    }
    lapStartDistanceM = lapEndDistanceM;
  }

  return weightedDistance > 0 ? weightedGap / weightedDistance : undefined;
}

export function gapPaceForSample(laps: ActivityLap[], sample: ActivitySample): number | undefined {
  if (typeof sample.distanceM !== "number" || !Number.isFinite(sample.distanceM)) {
    return undefined;
  }
  let lapStartDistanceM = 0;
  for (const lap of laps.slice().sort((left, right) => left.index - right.index)) {
    const lapDistanceM = finitePositiveOrZero(lap.distanceM);
    const lapEndDistanceM = lapStartDistanceM + lapDistanceM;
    if (sample.distanceM >= lapStartDistanceM && sample.distanceM <= lapEndDistanceM && validPace(lap.avgGradeAdjustedPaceSPKM)) {
      return lap.avgGradeAdjustedPaceSPKM;
    }
    lapStartDistanceM = lapEndDistanceM;
  }
  return undefined;
}

function samplesForClimb(samples: ActivitySample[], climb: ActivityClimb) {
  return samples
    .filter((sample) => sample.index >= climb.startSampleIndex && sample.index <= climb.endSampleIndex)
    .slice()
    .sort((left, right) => left.index - right.index);
}

function sampleDurationSeconds(previous: ActivitySample, current: ActivitySample, distanceDeltaM: number) {
  const elapsedDeltaS = difference(previous.elapsedS, current.elapsedS);
  if (elapsedDeltaS !== undefined) {
    return elapsedDeltaS;
  }

  if (previous.timestamp && current.timestamp) {
    const timestampDeltaS = (Date.parse(current.timestamp) - Date.parse(previous.timestamp)) / 1000;
    if (Number.isFinite(timestampDeltaS) && timestampDeltaS > 0) {
      return timestampDeltaS;
    }
  }

  const speedMPS = [current.speedMPS, previous.speedMPS].find((value) => validSpeed(value));
  return speedMPS === undefined ? undefined : distanceDeltaM / speedMPS;
}

function difference(previous?: number, current?: number) {
  if (typeof previous !== "number" || !Number.isFinite(previous) || typeof current !== "number" || !Number.isFinite(current)) {
    return undefined;
  }
  const result = current - previous;
  return result > 0 ? result : undefined;
}

function finitePositiveOrZero(value?: number) {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 ? value : 0;
}

function finiteDistance(value?: number) {
  return typeof value === "number" && Number.isFinite(value) && value >= 0 ? value : undefined;
}

function validSpeed(value?: number): value is number {
  return typeof value === "number" && Number.isFinite(value) && value > 0;
}

function validPace(value?: number): value is number {
  return typeof value === "number" && Number.isFinite(value) && value > 0;
}

File: internal/app/climb_performance.go
package app

import (
	"math"
	"sort"
)

func enrichActivityClimbPerformance(climbs []ActivityClimb, samples []ActivitySample, laps []ActivityLap) []ActivityClimb {
	for index := range climbs {
		climbs[index].PaceSPKM = climbPaceForSamples(samples, climbs[index])
		climbs[index].GapSPKM = climbGapForLaps(laps, climbs[index])
	}
	return climbs
}

func climbPaceForSamples(samples []ActivitySample, climb ActivityClimb) *float64 {
	climbSamples := make([]ActivitySample, 0)
	for _, sample := range samples {
		if sample.Index >= climb.StartSampleIndex && sample.Index <= climb.EndSampleIndex {
			climbSamples = append(climbSamples, sample)
		}
	}
	sort.SliceStable(climbSamples, func(left, right int) bool {
		return climbSamples[left].Index < climbSamples[right].Index
	})
	if len(climbSamples) < 2 {
		return nil
	}

	var distanceM, movingTimeS float64
	for index := 1; index < len(climbSamples); index++ {
		previous := climbSamples[index-1]
		current := climbSamples[index]
		if previous.DistanceM == nil || current.DistanceM == nil || !finiteClimbValue(*previous.DistanceM) || !finiteClimbValue(*current.DistanceM) || *previous.DistanceM < 0 || *current.DistanceM < 0 {
			continue
		}
		distanceDeltaM := *current.DistanceM - *previous.DistanceM
		if distanceDeltaM <= 0 {
			continue
		}

		durationS := climbSampleDurationSeconds(previous, current, distanceDeltaM)
		if durationS == nil || *durationS <= 0 {
			continue
		}
		distanceM += distanceDeltaM
		movingTimeS += *durationS
	}
	if distanceM <= 0 || movingTimeS <= 0 {
		return nil
	}
	paceSPKM := movingTimeS / distanceM * 1000
	return &paceSPKM
}

func climbGapForLaps(laps []ActivityLap, climb ActivityClimb) *float64 {
	startDistanceM := math.Min(climb.StartDistanceM, climb.EndDistanceM)
	endDistanceM := math.Max(climb.StartDistanceM, climb.EndDistanceM)
	if !finiteClimbValue(startDistanceM) || !finiteClimbValue(endDistanceM) {
		return nil
	}

	sortedLaps := append([]ActivityLap(nil), laps...)
	sort.SliceStable(sortedLaps, func(left, right int) bool {
		return sortedLaps[left].Index < sortedLaps[right].Index
	})
	var lapStartDistanceM, weightedGap, weightedDistance float64
	for _, lap := range sortedLaps {
		lapDistanceM := lap.DistanceM
		if !finiteClimbValue(lapDistanceM) || lapDistanceM < 0 {
			lapDistanceM = 0
		}
		lapEndDistanceM := lapStartDistanceM + lapDistanceM
		overlapDistanceM := math.Max(0, math.Min(endDistanceM, lapEndDistanceM)-math.Max(startDistanceM, lapStartDistanceM))
		if overlapDistanceM > 0 && validClimbPace(lap.AvgGradeAdjustedPaceSPKM) {
			weightedGap += *lap.AvgGradeAdjustedPaceSPKM * overlapDistanceM
			weightedDistance += overlapDistanceM
		}
		lapStartDistanceM = lapEndDistanceM
	}
	if weightedDistance <= 0 {
		return nil
	}
	gapSPKM := weightedGap / weightedDistance
	return &gapSPKM
}

func climbSampleDurationSeconds(previous, current ActivitySample, distanceDeltaM float64) *float64 {
	if elapsedS := positiveClimbDifference(previous.ElapsedS, current.ElapsedS); elapsedS != nil {
		return elapsedS
	}
	if previous.Timestamp != nil && current.Timestamp != nil {
		timestampDeltaS := current.Timestamp.Sub(*previous.Timestamp).Seconds()
		if timestampDeltaS > 0 && finiteClimbValue(timestampDeltaS) {
			return &timestampDeltaS
		}
	}
	for _, speedMPS := range []*float64{current.SpeedMPS, previous.SpeedMPS} {
		if speedMPS != nil && finiteClimbValue(*speedMPS) && *speedMPS > 0 {
			durationS := distanceDeltaM / *speedMPS
			return &durationS
		}
	}
	return nil
}

func positiveClimbDifference(previous, current *int) *float64 {
	if previous == nil || current == nil {
		return nil
	}
	difference := float64(*current - *previous)
	if difference <= 0 {
		return nil
	}
	return &difference
}

func validClimbPace(value *float64) bool {
	return value != nil && finiteClimbValue(*value) && *value > 0
}

func finiteClimbValue(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0)
}
